rfc: local git urls (#357)
kubecfg/kubecfg: docs/rfcs/rfc04.md

Each "cloud provider" and "deployment phase" have their own files where the common aspects for each those layers would be expressed.
If you want you can also factor out common things for each region, but I'm leaving this out from this document since it's irrelevant for the actual proposal.

Here's the import tree for the more visual minded of the readers (arrows means "imports"):

```mermaid
%%{init: { 'theme': 'default', 'flowchart': {'curve': 'linear', 'rankSpacing': 120, 'nodeSpacing': 20} } }%%
graph RL;
    aws(aws)
    prod(prod)
    app(app)
    stag(stag)
    gcp(gcp)
 
    prod-aws-us-east-1(us-east-1)-->prod-aws-common
    prod-aws-us-west-1(us-west-1)-->prod-aws-common
    prod-aws-common(common)-->aws & prod

    prod-gcp-us-east-1(us-east-1)-->prod-gcp-common
    prod-gcp-us-west-1(us-west-1)-->prod-gcp-common
    prod-gcp-common(common)-->gcp & prod

    stag-aws-us-east-1(us-east-1)-->stag-aws-common
    stag-aws-us-west-1(us-west-1)-->stag-aws-common
    stag-aws-common(common)-->aws & stag

    stag-gcp-us-east-1(us-east-1)-->stag-gcp-common
    stag-gcp-us-west-1(us-west-1)-->stag-gcp-common
    stag-gcp-common(common)-->gcp & stag

    prod-->app
    stag-->app

> ```

Author: I choose the graph orientation to mimic the file tree orientation; the arrows mean "imports" and they go to the left.

> ```

Author: ah perhaps you're saying that the mermaid source could be written using the other orientation

    prod-gcp-us-east-1(us-east-1)-->prod-gcp-common
    prod-gcp-us-west-1(us-west-1)-->prod-gcp-common
    prod-gcp-common(common)-->gcp & prod

    stag-aws-us-east-1(us-east-1)-->stag-aws-common
    stag-aws-us-west-1(us-west-1)-->stag-aws-common
    stag-aws-common(common)-->aws & stag

    stag-gcp-us-east-1(us-east-1)-->stag-gcp-common
    stag-gcp-us-west-1(us-west-1)-->stag-gcp-common
    stag-gcp-common(common)-->gcp & stag

    prod-->prod-artifact

    prod-->app
    stag-->app

    stag-->stag-artifact

    style prod-artifact fill:#8f8
    style stag-artifact fill:#8f8
```

The "production" template thus imports both the "app" template and the "v1" inventory JSON file. The "production" template overlays the "app" template with the "v1" inventory config, effectively causing every deployment that imports the staging template into a deployment that runs "v1" of the software.

## Problem

The main problem with this setup is that while each "phase" can have some specific overlays, and we make use of that feature to use a different version of the main application binaries in staging and in production, we have only one version of the main templates.

> > Any change to the templates applies immediately on every target cluster.

Reviewer: ```suggestion
> [!IMPORTANT]  
> Any change to the app templates applies immediately on every target cluster.
```

I think your intent was to call attention to this as a sort of "problem statement." This is Github-specific markdown, but would help with the callout. There are also "NOTE" and "WARNING" options.

![image](https://github.com/kubecfg/kubecfg/assets/2457559/ae6494db-dc67-4c7b-890b-e81fb5a729e4)

I've also clarified the templates to which you're referring.

1. It's hard to automate the merging of the overlay back in the original templates. This means the developer must remember to follow the rollout process and manually promote the change when it's safe to do so. That's toil and it's also brittle.

## Insights

1. The main insight is to recognize that the Jsonnet template **is a program**.

2. The other insight is to recognize that the cluster specific entrypoints are the program. The main application template they import is just **a dependency**.

   In the same vein as when you have your application A that depends on your library L at version V1, and another application B that depends on your library L at version V2, the templates for cluster A depend on the templates from library L at version V1 and cluster B depends on the templates from library L at version V2.

   Most programming languages have package managers that express dependencies that include name _and_ version. If we do the same with the dependency between the cluster templates and the _application templates_ we'd solve the [problem](#Problem).

> 3. The reason we don't normally think in terms of (1) and (2) is that the cluster-specific Jsonnet entry points are small and the bulk of the templates live in the _library_. Because of this it goes against the grain to fully split the _applications_ (cluster entrypoints like our `deploy/prod/aws/us-east-1.jsonnet`) from the _library_ (the `app/app.jsonnet`). It would be too many repositories to handle. It's too much work, and we're drawn to think of it in another way

Reviewer: ```suggestion
3. The reason we don't normally think in terms of (1) and (2) is that the cluster-specific Jsonnet entry points are small and the bulk of the templates live in the _library_. Because of this it goes against the grain to fully split the _applications_ (cluster entrypoints like our `deploy/prod/aws/us-east-1.jsonnet`) from the _library_ (the `app/app.jsonnet`) in separate repositories. It would be too many repositories to handle. It's too much work, and we're drawn to think of it in another way.
```

"_split how_" clarity for this sentence

> 4. The last insight is that we don't need to split in separate repos. All we need is a way to refer to old _revisions_ within the same repository.

Reviewer: This is a meta-insight that refers to insight (3) rather than an insight borne of the "Problem" above. Perhaps it belongs outside the list with context from the list baked-in?
```suggestion

We don't need to separate cluster-specific entrypoints and the library of app templates into repositories. All we need is a way to refer to old _revisions_ within the same repository.
```

Luckily Jsonnet an extensible offers import hooks functionality that kubecfg already taps extensively.

## Proposal

> We propose to add a `srcfs` URL scheme and an import hook implementation that will allow to import file from the current project's version control system as it was a specific revision.

Reviewer: ```suggestion
We propose to add a `srcfs` URL scheme and an import hook implementation that will allow to import files from the current project's version control system as it was a specific revision.
```

> We propose to add a `srcfs` URL scheme and an import hook implementation that will allow to import file from the current project's version control system as it was a specific revision.

Reviewer: It wasn't immediately clear to me what `srcfs` would do; I had to read your description here. 

I think you mean `src` as in "source control." If you intend to leave the implementation open to version control systems other than git, then perhaps this could be `vcsfs`?

This could be a problem that others won't share with me, so please speak out if I'm alone here.

Author: yeah; the reason the name is srcfs is a reference/allusion to google's internal tool; in pretty much the same way as kubecfg is also an allusion.

I think I'll move to something more explicit like `git+file:` or something `local-git:`

> This will allow you to _pin_ the version of the basic templates while still having a real template at hand. This differs from snapshotting a _rendered_ set of k8s manifests in that you can still pass parameters and apply further overlays that are meant to be applied _instantly_ without having to go through the promotion automation. This is useful when you need to perform GitOps operations under time constrains (e.g. during an incident response).

### Lifecycle

Let's see how this feature is meant to be used in practice to continuously deploy the latest tested version of an app to production.

Let's assume we have a given version of our application+config template bundle being deployed in staging and that we have some automated Post Deployment Tests telling us whether everything is healthy.

All we need to do is to make an automated PR that "updates" the `history-channel/prod` file to point to whatever is the current "healthy" version in staging.

> > I spent quite some time fighting with mermaid to draw this and I ultimately gave up and drew it by hand. So even if you're not a visual person, I'm sure you'll admire the primary colors of the following diagrams.

Reviewer: Excellent handwriting :wink:

A quick test on a production repo containing more than 1GB of git history:

```console
$ time git clone $REPO full
...
1m22.221s
$ time git clone --depth 1 $REPO shallow 
...
0m3.611s
$ time git clone --filter=tree:0 $REPO treeless
...
0m10.167s
$ du -sh full
1.4G	full
$ du -sh shallow
119M	shallow
$ du -sh treeless
144M	treeless
```

So it takes a few seconds more to clone a _treeless_ and you'll need to add
roughtly the same delta (3-6s) for each of the individual revisions you'd end
up lazily fetching. It's a bit slower than a shallow clone but it's a far cry from the cost of a full clone.

While the `git` CLI knows how to lazily fetch objects, not all Git clients do. The Go `go-git` library doesn't know how to do so we need to pre-fetch the commits.

An efficient way to ensure tree and blobs are all beeing lazily fetched is:

```console
$ git archive --format=tar --output=/dev/null $REVISION
```

### Vendoring

[RF03 Caching and Vendoring](https://github.com/kubecfg/kubecfg/blob/vendoring/docs/rfcs/rfc03.md) describes a general mechanism for vendoring external references.

> Should `srcfs` be considered an internal or an external reference?

Reviewer: ```suggestion
#### Should `srcfs` be considered an internal or an external reference?
```

I thought you were presenting an open question at first.
